examples/text/plot_document_clustering.py

```diff
@@ -3,284 +3,363 @@
+print(f"Explained variance of the SVD step: {explained_variance * 100:.1f}%")
 
-if opts.n_components:
-    print("Performing dimensionality reduction using LSA")
-    t0 = time()
-    # Since LSA/SVD results are not normalized,
-    # we redo the normalization to improve the k-means result.
-    svd = TruncatedSVD(opts.n_components)
-    normalizer = Normalizer(copy=False)
-    lsa = make_pipeline(svd, normalizer)
+# %%
+# We are now ready to use the `"k-means++"` init. As it requires a single
+# initialization, the processing time will be reduced for both
+# :class:`~sklearn.cluster.KMeans` and :class:`~sklearn.cluster.MiniBatchKMeans`.
+
+kmeans = KMeans(
+    n_clusters=true_k,
+    init="k-means++",
+    max_iter=100,
+    n_init=1,
+    random_state=0,
+)
 
-    X = lsa.fit_transform(X)
+fit_and_evaluate(kmeans, X, name="KMeans with LSA")
 
-    print("done in %fs" % (time() - t0))
+# %%
+# We repeat the experiment with :class:`~sklearn.cluster.MiniBatchKMeans`.
+
+minibatch_kmeans = MiniBatchKMeans(
+    n_clusters=true_k,
+    init="k-means++",
+    n_init=1,
+    init_size=1000,
+    batch_size=1000,
+    random_state=0,
+)
 
-    explained_variance = svd.explained_variance_ratio_.sum()
-    print(
-        "Explained variance of the SVD step: {}%".format(int(explained_variance * 100))
-    )
+fit_and_evaluate(minibatch_kmeans, X, name="MiniBatchKMeans\nwith LSA")
 
+# %%
+# **Top terms per cluster**
+#
+# Since :class:`~sklearn.feature_extraction.text.TfidfVectorizer` and can be
+# inverted we can identify the cluster centers, which provide an intuition of
+# the most predictive words **per cluster**. See the example script
+# :ref:`sphx_glr_auto_examples_text_plot_document_classification_20newsgroups.py`
+# for a comparison with the most predictive words **per class**.
+
+original_space_centroids = lsa[0].inverse_transform(kmeans.cluster_centers_)
+order_centroids = original_space_centroids.argsort()[:, ::-1]
+terms = vectorizer.get_feature_names_out()
+
+for i in range(true_k):
+    print(f"Cluster {i}: ", end="")
+    for ind in order_centroids[i, :10]:
+        print(f"{terms[ind]} ", end="")
     print()
 
+# %%
+# HashingVectorizer
+# -----------------
+# A similar experiment can be done using a
+# :class:`~sklearn.feature_extraction.text.HashingVectorizer` instance, which
+# does not provide IDF weighting as this is a stateless model (the fit method
+# does nothing). Here we illustrate the effect of Tfidf weighting by
+# benchmarking the function :class:`~sklearn.cluster.KMeans` with and without
+# such normalization.
+
+from sklearn.feature_extraction.text import HashingVectorizer
+
+hasher = HashingVectorizer(
+    stop_words="english",
+    alternate_sign=False,
+    norm=None,
+)
+
+t0 = time()
+X = hasher.fit_transform(dataset.data)
+print(f"vectorizing done in {time() - t0:.3f} s")
+print(f"n_samples: {X.shape[0]}, n_features: {X.shape[1]}")
+
+fit_and_evaluate(kmeans, X, name="KMeans with\nsimple hashing")
```

Comment:
Fitting `KMeans` with `init="kmeans++"` on very high dimensional data can randomly lead to very degenerate clustering and is therefore very unstable. I think we should only fit this model and the following on pipelines with the LSA strategy to avoid the problem.